### PR TITLE
653 progress bar for fluka

### LINKED ISSUE #653: Progress bar for Fluka

### PR DESCRIPTION
**Pull request by hendzeld**
Resolves 653

### CODE REVIEW

**Comment by grzanka** [comment]
Note that https://github.com/yaptide/yaptide/pull/575 was merged

**Comment by grzanka** [comment]
Still draft ?

File: yaptide/celery/utils/progress/fluka_monitor.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+                        varbose: bool = False) -> None:
+    """Function reading the fluka output file and raporting progress to the backend."""
+    in_progress = False
+    requested_primaries = 0
+    update_time = 0
+    for line in linte_iterator:
+        if event.is_set():
+            return
+        utc_now = time_now_utc()
+        if re.match(S_OK_OUT_START, line):
```
Comment (grzanka) [inline yaptide/celery/utils/progress/fluka_monitor.py]:
```suggestion
        if re.match(S_OK_OUT_START, line):
```

I suspect that matching lines based on regexp may be costly. What about using `startstwith` method of a string ?
I think it should be worth to examine what is faster.

Another thing is that the order of regexp in the loop is important. If you have a line which you expect to be in the first part of the file and will appear only once, put it at the end of the loop. First should be the lines that you expect to be most frequent.

File: yaptide/celery/utils/pymc.py [hunk 2]
```diff
@@ -293,6 +294,59 @@ def read_file(event: threading.Event,
     send_task_update(simulation_id, task_id, update_key, up_dict)
 
 
+def read_fluka_file(event: threading.Event,
+                    dirpath: Path,
+                    simulation_id: int,
+                    task_id: str,
+                    update_key: str,
+                    timeout_wait_for_file: int = 20,
+                    timeout_wait_for_line: int = 5 * 60,
```
Comment (grzanka) [inline yaptide/celery/utils/pymc.py]: Can you add `_s` suffix to the variable name to denote which time unit we deal with ?

File: yaptide/celery/utils/pymc.py [hunk 3]
```diff
@@ -81,7 +82,7 @@ def command_to_run_fluka(dir_path: Path, task_id: str) -> list[str]:
         input_path=dir_path,  # skipcq: PYL-W0612 # usefull
         simulator_type=SimulatorType.fluka,
         simulator_exec_path=None,  # useless, we guess from PATH
-        cmdline_opts="")  # useless, not useful for fluka
+        cmdline_opts="-M 1")  # useless, not useful for fluka
```
Comment (grzanka) [inline yaptide/celery/utils/pymc.py]: why this option ?

Comment (hendzeld) [reply yaptide/celery/utils/pymc.py]: I will add an appropriate comment.

About this option:
Fluka by default executes  5 simulations one after another.
From fluka -h (help) sommand
```
  This operation is repeated from the cycle after the one indicated
       on the -N option (default 0) to the one indicated on the -M
       option (default 5).
...

       -N<last_cycle>   number of the previous run cycle number
       -M<final_cycle>  final run number
```

Comment (grzanka) [reply yaptide/celery/utils/pymc.py]: Add a comment please. Usually we run multiple cycles when working in command line. That gives us error estimation when running on single CPU

File: yaptide/celery/tasks.py [hunk 4]
```diff
@@ -273,3 +286,29 @@ def monitor_shieldhit(event: threading.Event, tmp_work_dir: str, task_id: int, u
 
     logging.info("No monitoring processes started for task %d", task_id)
     return None
+
+
+def monitor_fluka(event: threading.Event, tmp_work_dir: str, task_id: int, update_key: str,
+                  simulation_id: int) -> Optional[MonitorTask]:
+    """Function running the monitoring process for SHIELDHIT simulation"""
```
Comment (grzanka) [inline yaptide/celery/tasks.py]:
```suggestion
    """Function running the monitoring process for Fluka simulation"""
```

File: yaptide/celery/utils/progress/fluka_monitor.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+S_OK_OUT_INIT = "Total time used for initialization:"
+S_OK_OUT_START = "1NUMBER OF BEAM"
+S_OK_OUT_IN_PROGRESS = " NEXT SEEDS:"
+S_OK_OUT_COLLECTED = " All cases handled by Feeder"
+S_OK_OUT_FIN_PATTERN = re.compile(r"^ \* ======(?:( )*?)End of FLUKA [\w\-.]* run (?:( )*?) ====== \*")
+
+logger = logging.getLogger(__name__)
+
+
+def parse_progress_remaining_line(line: str) -> Optional[Tuple[int, int]]:
```
Comment (grzanka) [inline yaptide/celery/utils/progress/fluka_monitor.py]:
```suggestion
def parse_progress_remaining_line(line: str) -> Optional[tuple[int, int]]:
```

I think we can stick to python > 3.9

**Review by grzanka — CHANGES_REQUESTED** [review]
fix deepsource issues

**Review by grzanka — APPROVED** [review]

**Pull request merged**